contents/data/nosql/NoSQL.md

```diff
@@ -0,0 +1,437 @@
+  }
+}
+```
+
+<br>
+
+### 3. Straightforward Expression of Relations in Data
+
+The world is full of relations. For example, a patient _is related to_ her disease record,
+just as a customer _is related to_ their shopping cart.
+
+Sometimes, the objects on both sides of the relation can contain substantial amounts of
+information, and may be impossible to store as a singular field in one or the other document.
+
+Hence, simple relations such as
+<tooltip content="A type of relation where each item is only related to one other item">
+_one-to-one_
+</tooltip>
+relations,
+<tooltip content="A type of relation where each item can be related to many other items, but these other items are only ever related to one item on the other end">
+_one-to-many_
+</tooltip>
+relations are often expressed in document databases simply in the form of embedded documents, which is made possible due to 
+the schema-less characteristic of NoSQL databases.
+
+For example, for a customer and his / her shopping cart, we may have the following:
+```js
+{
+  username: "panda",
+  cart: {
+    totalPrice: 100,
+    cartItems: [ ... ],
+    discountCode: "panda"
+  },
+  email: "panda@pandas.com",
+  ...
+}
+```
+
+In the case of more complicated
+<tooltip content="A type of relation where each item can be related to many other items, and these other items can also be related to many items on the other end">
+_many-to-many_
+</tooltip>
+relationships, relations are commonly stored using references, to avoid duplication of data.
+
+For example, items in an e-commerce website are related to the many customers through their carts.
+In these carts, it is much more space efficient to store references to the items, than the item documents themselves.
+
+In this example, the uniquely generated `_id` field for each item document in the items collection could be one such _reference_:
+```js
+{
+  type: "book",
+  price: 20,
+  popularity: 9.7,
+  _id: "9d1793bd491349n913847n93d"
+}
+```
+
+In the user's cart, we would simply store these `_id` _references_, which are used to lookup the item documents in the items collection later:
+```js
+cart: {
+  totalPrice: 100,
+  cartItems: [
+    "9d1793bd491349n913847n93d",
+    "9d1793bd491349n913847njh8",
+  ],
+  discountCode: "panda"
+}
+```
+
+
+
+<box type="info" light>
+For this reason, many NoSQL database solutions (e.g. MongoDB) implement a unique <code>id</code> field
+for each document by default.
+</box>
+
+---
+
+## Why NoSQL?
+
+<br>
+
+### 1. Highly Suited for Iterated Development
+Although less mature than SQL databases, NoSQL databases were designed to solve many of the emerging challenges in databases today.
+
+One of the most consequential impacts NoSQL has had was enabling faster iterated development.
+Given the highly flexible relational structure of NoSQL databases, and the schemaless format of documents 
+in NoSQL, this means that developers can adapt the database quicker to changing customer and business requirements.
```

Comment:
Should we mention the fact that in contrast SQL database are harder to change later due to the difficulty in modifying the schema?